[Refactor] Change `PokemonSummonData` to an interface

DayKev opened the pull request:
## What are the changes the user will see?
Transformed Pokémon will stay transformed when loading save data.

## Why am I making these changes?
Code cleanup.
Fixes a bug with Imposter/Transform.

## What are the changes from a developer perspective?
- `PokemonSummonData` changed from a class to an interface.
- `Pokemon#summonData` is now always initialized in the `Pokemon` class constructor via `resetSummonData()`.
- `PokemonData` constructor refactored/simplified.
- Removed `null` checks for `summonData`.
- Various tests updated.

## How to test the changes?
`npm run test:silent`

## Checklist
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test:silent`)
  - [x] Have I created new automated tests (`npm run test:create`) or updated existing tests related to the PR's changes?

DayKev committed:
[Refactor] Change `PokemonSummonData` to an interface

`Pokemon#summonData` is now always initialized
in the `Pokemon` class constructor

Simplify `PokemonData` constructor

Fix `stockpiling.test.ts`

DayKev committed:
Remove `null` checks for `summonData`

DayKev committed:
Fix Imposter and Transform tests

DayKev committed:
Fix Imprison tests

DayKev committed:
Move `PokemonSummonData` to `src/@types/PokemonSummonData.ts`

DayKev committed:
Revert `summonData.addedType` to `| null`

DayKev committed:
Fix Transform and Imposter not being loaded from save data

flx-sta commented on src/data/custom-pokemon-data.ts:
```diff
@@ -6,6 +6,7 @@ import type { Nature } from "#enums/nature";
 /**
  * Data that can customize a Pokemon in non-standard ways from its Species
  * Currently only used by Mystery Encounters and Mints.
+ * @todo Make this an interface
```
Already preparing the next PR, are you ? :D

DayKev replied:
Yep, I'm on a quest to remove all these unnecessary classes. 🔥

flx-sta commented on src/data/species-form-change-triggers/species-form-change-post-move-trigger.ts:
```diff
@@ -3,6 +3,6 @@ import type { Pokemon } from "#app/field/pokemon";
 export class SpeciesFormChangePostMoveTrigger extends SpeciesFormChangeMoveTrigger {
   override canChange(pokemon: Pokemon): boolean {
-    return pokemon.summonData && pokemon.getLastXMoves(1).some((m) => this.movePredicate(m.move.id)) === this.used;
+    return pokemon.getLastXMoves(1).some((m) => this.movePredicate(m.move.id)) === this.used;
```
I'm wondering if we should have an alias called `getLatestMove()` as the most frequent use-case of `getLastXMoves` is with `1` as a param

<sup>Just sharing my thoughts, feel free to ignore</sup>

DayKev replied:
Yeah, adding a helper for this might be good to prevent mistakes (there's a few places where this and/or `getMoveHistory` are used incorrectly when trying to get a Pokémon's last used move).

DayKev replied:
Created an issue for it #1155

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -1038,12 +1038,10 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+      this.summonData.statStages[stat - 1] = Math.min(value, 6);
+    } else {
+      this.summonData.statStages[stat - 1] = Math.max(value, -6);
     }
```
I just noticed this can be solved with `Phaser.Math.Clamp` which makes the `if` obsolete:

```suggestion
   this.summonData.statStages[stat - 1] = Phaser.Math.Clamp(value, -6, 6);
```

https://docs.phaser.io/phaser/concepts/math#clamp

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -1350,25 +1348,25 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
 
   getMoveset(baseOnly?: boolean): PokemonMove[] {
-    const ret = !baseOnly && this.summonData?.moveset ? this.summonData.moveset : this.moveset;
+    const ret = !baseOnly && this.summonData.moveset.length ? this.summonData.moveset : this.moveset;
```
NIT: Not length check shortcuts

```suggestion
    const ret = !baseOnly && this.summonData.moveset.length !== 0 ? this.summonData.moveset : this.moveset;
```

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -1350,25 +1348,25 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     }
-
-    return ret;
+    if (!this.isPlayer()) {
```
NIT: Use `isEnemy()` instead? Or is that not the same?

```suggestion
    if (this.isEnemy()) {
```

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -1669,7 +1667,7 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
       return false;
     }
-    if (this.summonData?.abilitySuppressed && !ability.hasAttrFlag(AbAttrFlag.UNSUPPRESSABLE_ABILITY)) {
+    if (this.summonData.abilitySuppressed && !ability.hasAttrFlag(AbAttrFlag.UNSUPPRESSABLE_ABILITY)) {
```
Here is a typo I noticed.
It's not unsuppress~~able~~ but unsuppress**ible**

<sub>If this exceeds your PRs scope just let me know</sub>

DayKev replied:
I think both might be valid.

https://www.oed.com/dictionary/unsuppressible_adj
https://www.oed.com/dictionary/unsuppressable_adj

https://en.wiktionary.org/wiki/unsuppressible
https://en.wiktionary.org/wiki/unsuppressable
> Alternative form of [unsuppressible](https://en.wiktionary.org/wiki/unsuppressible#English)

flx-sta replied:
Interesting. My new spellchecker extension doesn’t like the “able” one

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -2212,7 +2210,8 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     this.moveset[moveIndex] = move;
-    if (this.summonData?.moveset) {
+    // TODO: should this also be modifying the summon data moveset?
+    if (this.summonData.moveset.length) {
```
Not length check shortcuts please

```suggestion
    if (this.summonData.moveset.length !== 0) {
```

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -2665,12 +2664,8 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-      return ret;
-    }
-    return null;
+  getOpponent(targetIndex: number): Pokemon | undefined {
```
NIT: tsdocs

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -3535,10 +3530,7 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-    if (!this.summonData) {
-      return null;
-    }
+  getTag<T extends BattlerTag = BattlerTag>(...tagTypes: BattlerTagType[]): T | undefined {
```
NIT: tsdocs

DayKev replied:
There's some cleanup/refactoring I want to do around `BattlerTag`s similar to what I did with `ArenaTag`s and adding TSDocs to `BattlerTag`-related functions is more in scope for that than this PR.

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -3548,30 +3540,18 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-    if (!this.summonData) {
-      return null;
-    }
+  findTag<T extends BattlerTag = BattlerTag>(tagFilter: (tag: BattlerTag) => boolean): T | undefined {
```
NIT: tsdocs

flx-sta replied:
Same here I assume

flx-sta commented on src/field/pokemon.ts:
```diff
@@ -4161,15 +4129,26 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
   }
 
   resetSummonData(): void {
```
NIT: tsdocs

flx-sta commented on src/modifier/modifier.ts:
```diff
@@ -1649,8 +1649,7 @@ export class FlinchChanceModifier extends PokemonHeldItemModifier {
   override apply(pokemon: Pokemon, flinched: BooleanHolder): boolean {
-    // The check for pokemon.summonData is to ensure that a crash doesn't occur when a Pokemon with King's Rock procs a flinch
-    if (pokemon.summonData && !flinched.value && pokemon.randSeedInt(100) < this.getStackCount() * this.chance) {
+    if (!flinched.value && pokemon.randSeedInt(100) < this.getStackCount() * this.chance) {
```
What the? Why does `Pokemon` have it's own `randSeedInt()` method attached?

DayKev replied:
If `Pokemon#randSeedInt` is called it uses the current battle's seed (if possible) instead of the global seed.
Yeah, it's kinda jank; the way RNG is done needs to be refactored.

flx-sta commented on src/phases/summon-phase.ts:
```diff
@@ -196,6 +196,10 @@ export class SummonPhase extends PartyMemberPokemonPhase {
                 pokemon.cry(pokemon.getHpRatio() > 0.25 ? undefined : { rate: 0.85 });
                 pokemon.getSprite().clearTint();
                 pokemon.resetSummonData();
+                // necessary to stay transformed during wild waves
```
What's a "Wild Wave"? A non trainer wave?

DayKev committed:
Apply suggestions

flx-sta commented on src/data/moves/move-attrs/last-move-double-power-attr.ts:
```diff
@@ -23,7 +23,7 @@ export class LastMoveDoublePowerAttr extends VariablePowerAttr {
 
   override apply(user: Pokemon, _target: Pokemon, _move: Move, power: NumberHolder): boolean {
-    const target = user.getOpponent(0);
+    const target = user.getOpponents()[0];
```
Alternatively:
```suggestion
    const [target] = user.getOpponents();
```

flx-sta commented:
@DayKev Conflicts to resolve

DayKev committed:
Apply suggestion to `last-move-double-power-attr.ts`

PigeonBar commented on src/phases/terastallization-phase.ts:
```diff
@@ -36,7 +37,7 @@ export class TerastallizationPhase extends BattlePhase {
-    if (this.pokemon.summonData) {
+    if (!isNil(this.pokemon.summonData.addedType)) {
       this.pokemon.summonData.addedType = null;
     }
```
```suggestion
    this.pokemon.summonData.addedType = null;
```

DayKev committed:
Merge branch 'beta' into refactor/summon-data-interface

PigeonBar commented on src/system/pokemon-data.ts:
```diff
@@ -102,39 +104,35 @@ export default class PokemonData {
+        // @ts-expect-error - `_formIndex` is protected but we can't use `.formIndex` because it's a getter and the class data is lost
+        source.summonData.speciesForm._formIndex,
+      );
+    }
```
```suggestion
    }
    for (const turnMove of this.summonData.moveHistory) {
      // This is required because the full class object doesn't exist in save data
      turnMove.move = allMoves.get(turnMove.move.id);
    }
    for (const turnMove of this.summonData.moveQueue) {
      // This is required because the full class object doesn't exist in save data
      turnMove.move = allMoves.get(turnMove.move.id);
    }
```

PigeonBar commented on src/field/pokemon.ts:
```diff
@@ -984,7 +984,7 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
    */
   getStats(bypassSummonData: boolean = true): number[] {
-    if (!bypassSummonData && this.summonData?.stats) {
+    if (!bypassSummonData && this.summonData.stats) {
```
The "empty" value for `summonData.stats` is an array of 0's, which is truthy, which would cause this function to return the same array of 0's. This `getStats` function is unused, so we could probably handle this by just deleting the function.

DayKev committed:
Apply suggestions

- Remove unused method `Pokemon#getStats`

- Remove unnecessary `if` in `terastallization-phase.ts`

- Add save data reconstruction for
`summonData.moveHistory` and `summonData.moveQueue`

DayKev committed:
Merge branch 'beta' into refactor/summon-data-interface

innerthunder commented on src/@types/pokemon-summon-data.ts:
```diff
@@ -0,0 +1,53 @@
+ * Currently all fields in {@linkcode Pokemon.summonData} are stored in save data.
+ * @todo Rework save data to be more efficient
+ */
+export interface PokemonSummonData {
```
NIT: Every field should have docs

innerthunder commented on src/data/moves/move-attrs/last-move-double-power-attr.ts:
```diff
@@ -23,7 +23,7 @@ export class LastMoveDoublePowerAttr extends VariablePowerAttr {
   }
 
   override apply(user: Pokemon, _target: Pokemon, _move: Move, power: NumberHolder): boolean {
```
Oof, this can probably be simplified quite a bit by looking at each Pokemon's `order` instead...

innerthunder commented on src/field/pokemon.ts:
```diff
@@ -780,7 +780,7 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
   }
 
   getSpeciesForm(ignoreOverride?: boolean): PokemonSpeciesForm {
```
NIT: `ignoreOverride` should be renamed to `bypassSummonData` for parity with other `Pokemon` methods with similar logic.

innerthunder commented on src/field/pokemon.ts:
```diff
@@ -1329,7 +1315,7 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
   }
 
   getGender(ignoreOverride?: boolean): Gender {
```
NIT: rename `ignoreOverride`

innerthunder commented on src/field/pokemon.ts:
```diff
@@ -1355,25 +1341,25 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
   abstract getBossSegmentIndex(): number;
 
   getMoveset(baseOnly?: boolean): PokemonMove[] {
```
NIT: can probably rename `baseOnly` to `bypassSummonData`

innerthunder commented on src/field/pokemon.ts:
```diff
@@ -1439,7 +1425,7 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
 
     if (!types.length || !includeTeraType) {
-      if (!baseOnly && this.summonData?.types && this.summonData.types.length > 0) {
+      if (!baseOnly && this.summonData.types.length > 0) {
```
NIT: another `baseOnly` vs `bypassSummonData` vs `ignoreOverride` inconsistency... I think you get the idea at this point. One of these names should be used for all methods.

innerthunder commented on src/field/pokemon.ts:
```diff
@@ -2217,7 +2203,8 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     const move = new PokemonMove(moveId);
     this.moveset[moveIndex] = move;
-    if (this.summonData?.moveset) {
+    // TODO: should this also be modifying the summon data moveset?
+    if (this.summonData.moveset.length > 0) {
```
So I checked this method (`setMove`) for references and didn't find anything... Is this method even used anywhere?

Tempo-anon replied:
It's used in `sketch-attr.ts` and `learn-move-phase.ts`

innerthunder replied:
Ah, so it's only used for permanent moveset changes. The changes to `summonData.moveset` should be removed, then, right?

<sub>This method also needs docs...</sub>

DayKev replied:
Maybe... what happens in mainline if a transformed pokemon levels up and should learn a move?

DayKev replied:
Meh, I guess it's such a niche interaction we can just ignore it.

Tempo-anon replied:
I think/remember transformed Pokemon just don't learn moves when leveling up. Since the only users of Transform normally are ditto and mew and I don't recall them ever attempting to learn new moves when transformed

DayKev replied:
Moro tested and confirmed that it learns the move normally and doesn't affect the transformed moveset.

innerthunder commented on src/field/pokemon.ts:
```diff
@@ -2670,14 +2657,10 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
-
+  /**
+   * @returns An array containing the active, on field pokemon on the team opposing this pokemon.
+   * @todo Add `| undefined` to return type hint
```
Why? An empty array is a valid output as is.

DayKev replied:
Yeah but there's nothing making sure you account for that, which has led to many crashes in the past when trying to call functions from an empty array.

DayKev committed:
Add some TSDocs to `PokemonSummonData` fields

DayKev committed:
Rename `ignoreOverride` and `baseOnly` params to `bypassSummonData`

DayKev committed:
Remove summon data modification in `Pokemon#setMove`
